src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs

```diff
@@ -56,27 +57,77 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
+        where TFilterType : IRouteHandlerFilter
+    {
+        // We have a RouteHandlerBuiler and GroupRouteBuilder-specific AddFilter methods for convenience so you don't have to specify both arguments most the time.
+        return builder.AddRouteHandlerFilter<RouteGroupBuilder, TFilterType>();
+    }
+
     /// <summary>
     /// Registers a filter given a delegate onto the route handler.
     /// </summary>
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
-    /// <param name="routeHandlerFilter">A <see cref="Delegate"/> representing the core logic of the filter.</param>
+    /// <param name="routeHandlerFilter">A method representing the core logic of the filter.</param>
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
-    public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, Func<RouteHandlerInvocationContext, RouteHandlerFilterDelegate, ValueTask<object?>> routeHandlerFilter)
+    public static TBuilder AddRouteHandlerFilter<TBuilder>(this TBuilder builder, Func<RouteHandlerInvocationContext, RouteHandlerFilterDelegate, ValueTask<object?>> routeHandlerFilter)
+        where TBuilder : IEndpointConventionBuilder
     {
-        builder.RouteHandlerFilterFactories.Add((routeHandlerContext, next) => (context) => routeHandlerFilter(context, next));
-        return builder;
+        return builder.AddRouteHandlerFilter((routeHandlerContext, next) => (context) => routeHandlerFilter(context, next));
     }
 
     /// <summary>
     /// Register a filter given a delegate representing the filter factory.
     /// </summary>
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
-    /// <param name="filterFactory">A <see cref="Delegate"/> representing the logic for constructing the filter.</param>
+    /// <param name="filterFactory">A method representing the logic for constructing the filter.</param>
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
-    public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate> filterFactory)
+    public static TBuilder AddRouteHandlerFilter<TBuilder>(this TBuilder builder, Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate> filterFactory)
+        where TBuilder : IEndpointConventionBuilder
     {
-        builder.RouteHandlerFilterFactories.Add(filterFactory);
+        builder.Add(endpointBuilder =>
+        {
+            // Even if we split RouteFilterMetadata across multiple metadata objects, RouteEndointDataSource will still use all of them.
+            // We're just trying to be good citizens and keep the Endpoint.Metadata as small as possible.
+            RouteFilterMetadata? filterMetadata = null;
```

Comment:
Nice! I like that this has the side-effect of making filters registered on an endpoint inspectable.

Reply:
I agree that it's a lot easier to inspect with a debugger this way.

I did make `RouteFilterMetadata` internal for now, so it's not really inspectable programmatically unless we decide to make the metadata public API. The nice thing is we can choose to do that whenever we want without breaking anyone. It doesn't even have to stay `RouteFilterMetadata`.